geoana/em/static/halfspace.py

```diff
@@ -0,0 +1,733 @@
+        return e
+
+    def current_density(self, xyz):
+        """Current density for a point current in a halfspace.
+
+       This method computes the current density for the point current in a halfspace at
+        the set of gridded xyz locations provided. Where :math:`\\rho`
+        is the electric resistivity and :math:`\\mathbf{E}` is the electric field
+        for the point current.
+        The current density :math:`\\mathbf{J}` is:
+
+       .. math::
+
+            \\mathbf{J} = \\frac{\\mathbf{E}}{\\rho}
+
+        Parameters
+        ----------
+        xyz : (..., 3) numpy.ndarray
+            Locations to evaluate at in units m.
+
+        Returns
+        -------
+        J : (..., 3) np.ndarray
+            Current density of point current in units :math:`\\frac{A}{m^2}`.
+
+        Examples
+        --------
+        Here, we define a point current with current=1A in a halfspace and plot the current density.
+
+        >>> import numpy as np
+        >>> import matplotlib.pyplot as plt
+        >>> from geoana.em.static import PointCurrentHalfSpace
+
+        Define the point current.
+
+        >>> rho = 1.0
+        >>> current = 1.0
+        >>> simulation = PointCurrentHalfSpace(
+        >>>     current=current, rho=rho, location=None
+        >>> )
+
+        Now we create a set of gridded locations and compute the current density.
+
+        >>> X, Y = np.meshgrid(np.linspace(-1, 1, 20), np.linspace(-1, 1, 20))
+        >>> Z = np.zeros_like(X)
+        >>> xyz = np.stack((X, Y, Z), axis=-1)
+        >>> j = simulation.current_density(xyz)
+
+        Finally, we plot the curent density.
+
+        >>> j_amp = np.linalg.norm(j, axis=-1)
+        >>> plt.pcolor(X, Y, j_amp, shading='auto')
+        >>> cb1 = plt.colorbar()
+        >>> cb1.set_label(label= 'Current Density ($A/m^2$)')
+        >>> plt.ylabel('Y coordinate ($m$)')
+        >>> plt.xlabel('X coordinate ($m$)')
+        >>> plt.title('Current Density for a Point Current in a Halfspace')
+        >>> plt.show()
+        """
+
+        j = self.electric_field(xyz) / self.rho
+        return j
+
+
+class DipoleHalfSpace:
+    """Class for a dipole in a halfspace.
+
+        The ``DipoleHalfSpace`` class is used to analytically compute the
+        potentials, current densities and electric fields within a halfspace due to a dipole source.
+        Surface is assumed to be at z=0.
+
+        Parameters
+        ----------
+        current : float
+            Electrical current in the point current (A). Default is 1A.
+        rho : float
+            Resistivity in the point current (:math:`\\Omega \\cdot m`).
+        location_a : array_like
+            Location of the A current electrode (m). Default is (-1, 0, 0).
+        location_b : array_like
+            Location of the B current electrode (m). Default is (1, 0, 0).
+        """
+
+    def __init__(self, rho, location_a=np.r_[-1, 0, 0], location_b=np.r_[1, 0, 0], current=1.0):
```

Comment:
for the same reasons as before, you need to have `location_a=None` and `location_b=None` as the default for those two kwargs, (and then proceed to set them as `[-1, 0, 0]` and `[1, 0, 0]` as you've done in the other classes.